### PR TITLE
Fix how we handle niche optimization with ZST

### LINKED ISSUE #95: Statement assignment assert failing
There are two cases where the assert in gotoc statement assignment is failing [0]

1. When the types are wrapped (e.g. in a Cell) array types when one has an explicit length (e.g., `[T;N]`) and the other is inferred (`[T;_]`). E.g.: `StructTag("tag-std::cell::RefCell<[bool; 16]>")` and `StructTag("tag-std::cell::RefCell<[bool; _]>")`

2. When the types are closures. E.g., `Pointer { typ: StructTag("tag-[closure@file.rs:536:13: 546:14]") }` and `Pointer { typ: StructTag("tag-[closure@file.rs:503:13: 509:14]") }`

The root cause is the naming of these symbol associated with these types.

[0] https://github.com/model-checking/rmc/blob/main-151-2021-04-16/compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/stmt.rs#L153

### LINKED ISSUE #277: Fix casting warnings in `Projection::new`
There is a commented-out assertion with a TODO in `Projection::new`: the assertion fails on some regression tests; but also on some standard library issues that would be useful to have an actual check for. Once we fix the regressions we should enable it.

```rust
        // TODO: these assertions fail on a few regressions. Figure out why.
        // I think it may have to do with boxed fat pointers.
        // assert!(
        //     Self::check_expr_typ(&goto_expr, &mir_typ_or_variant, ctx),
        //     "\n{:?}\n{:?}",
        //     &goto_expr,
        //     &mir_typ_or_variant
        // );
```

### LINKED ISSUE #729: Error "Reached assignment statement with unequal types Pointer" with Result<(), Error>
I tried this code:

```rust
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0 OR MIT

enum Error {
    Error1,
    Error2,
}

fn to_option<T: Copy, E>(result: &Result<T, E>) -> Option<T> {
    if let Ok(v) = result {
        Some(*v)
    } else {
        None
    }
}

fn main() {
    let result: Result<(), Error> = Ok(());
    assert!(to_option(&result).is_some());
}

```

using the following command line invocation:

```
rmc <test_file>
```

with RMC version:

I expected to see this happen: Verification should succeed

Instead, this happened:
```
$ rmc fixme_niche_opt.rs | grep FAIL
(standard input):34:[assertion.1] Reached assignment statement with unequal types Pointer { typ: StructTag("tag-Unit") } Pointer { typ: StructTag("tag-_3943305294634710273") }: FAILURE
(standard input):66:VERIFICATION FAILED
```

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

The compiler uses niche optimization when there is only one variant that contains nonzero sized fields. This variant is represented as the dataful variant. However, other variants may have one or more ZST fields that can be deconstructed and referenced in the code.
    
Before this change, we were not generating code for them and it generates type mismatch issues when the user tried to access them. In the testcase I added, it also uncovered an issue that would crash the compiler when the names of the fields of the ZST variant did not match the dataful one.
    
This change ensures that we represent all variants that have any field, so their access is valid.

### Resolved issues:

Resolves #95.
Resolves #277.
Resolves #729 

### Call-outs:

I enable a `debug_assert` inside projection mismatch for now. I am planning to turn this into an assert and revert the logic added by https://github.com/model-checking/kani/pull/1057. But this is a much larger change that I prefer creating a separate PR for.

### Testing:

* How is this change tested? New test + fixme test.

* Is this a refactor change? Nope

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 1]
```diff
@@ -1211,32 +1228,37 @@ impl<'tcx> GotocCtx<'tcx> {
...
-            def.variants()
-                .iter_enumerated()
-                .map(|(i, case)| {
-                    Type::datatype_component(
+    ) -> Vec<DatatypeComponent> {
+        def.variants()
+            .iter_enumerated()
+            .filter_map(|(i, case)| {
+                if case.fields.is_empty() {
+                    // Skip variant types that cannot be referenced.
```
Comment (zhassan-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1246]: Nice!

**Review by zhassan-aws — APPROVED** [review]
Nice to finally get rid of this warning! Thanks!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 2]
```diff
@@ -1098,31 +1098,48 @@ impl<'tcx> GotocCtx<'tcx> {
                             }
                             fields.push(Type::datatype_component(
                                 "cases",
-                                ctx.codegen_enum_cases_union(
-                                    name,
-                                    adtdef,
-                                    subst,
-                                    variants,
-                                    initial_offset,
+                                ctx.ensure_union(
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:1101]: Sure!

**Comment by tedinski** [comment]
This PR rules. Just wanted to say that. Nice!

**Pull request merged**